entity.go

```diff
@@ -77,6 +77,38 @@ func NewMultipart(header Header, parts []*Entity) (*Entity, error) {
 	return New(header, r)
 }
 
+const defaultMaxHeaderBytes = 1 << 20 // 1 MB
+
+// ReadOptions represents the options of a message reader
+type ReadOptions struct {
+	// MaxHeaderBytes is the number of bytes
+	// the reader is allowed to read from a message.
+	// If zero, defaultMaxHeaderBytes is used.
+	MaxHeaderBytes int
+}
+
+// ReadWithOptions reads a message from r. ReadOptions is used to
+// manipulate header size limit.
+func ReadWithOptions(r io.Reader, options *ReadOptions) (*Entity, error) {
+	// set default options
+	if options == nil {
+		options = &ReadOptions{}
+	}
+	if options.MaxHeaderBytes == 0 {
+		options.MaxHeaderBytes = defaultMaxHeaderBytes
+	}
+
+	// use options.MaxHeaderBytes to set initial buffer size
+	// utilize this value in ReadHeader to validate header size
+	br := bufio.NewReaderSize(r, options.MaxHeaderBytes)
```

Comment:
This doesn't work. `bufio`'s buffer size doesn't limit the amount of bytes read, it limits the amount of bytes stored in memory.

Looking at it again, there may not be a nice way to limit the amount of bytes read.

Reply:
Yes, it was kind of a work around to check header size without providing additional parameters to `ReadHeader`. 

I had problem with `io.LimitedReader` about returning error when bytes read hit the limit. And using it to pass the test cases wasn't as nice looking as I expected at first like you mentioned. 

Also I am okay with closing the PR to not take more time since there are other guys waiting for a solution on https://github.com/emersion/go-message/issues/110 and someone else might have a quicker and better solution.

Reply:
No worries! I've opened https://github.com/emersion/go-message/pull/123 which should make it easy to add the reader options.

Reply:
Thanks for your time @emersion @foxcpp